torch/distributions/multivariate_normal.py

```diff
@@ -0,0 +1,154 @@
+    """
+    Performs a batched matrix-vector product, with an arbitrary batch shape.
+    """
+    batch_shape = bvec.shape[:-1]
+    event_dim = bvec.shape[-1]
+    bmat = bmat.expand(batch_shape + (event_dim, event_dim))
+    if batch_shape != bmat.shape[:-2]:
+        raise ValueError("Batch shapes do not match: matrix {}, vector {}".format(bmat.shape, bvec.shape))
+    bvec = bvec.unsqueeze(-1)
+    
+    # conform with `torch.bmm` interface, for matrices with `.dim() == 3`
+    if bvec.dim() == 2:
+        bvec.unsqueeze(0) #_
+     # flatten batch dimensions
+    bvec = bvec.contiguous().view((-1, event_dim, 1))
+    bmat = bmat.contiguous().view((-1, event_dim, event_dim)).expand((bvec.shape[0], -1, -1))
+    return torch.bmm(bmat, bvec).squeeze(-1).view(batch_shape+(event_dim,)) 
+
+
+def _batch_potrf_lower(bmat):
+    """
+    Applies a Cholesky decomposition to all matrices in a batch of arbitrary shape.
+    """
+    n = bmat.size(-1)
+    cholesky = torch.stack([C.potrf(upper=False) for C in bmat.unsqueeze(0).contiguous().view((-1,n,n))])
+    return cholesky.view(bmat.shape)
+
+
+def _batch_diag(bmat):
+    """
+    Returns the diagonals of a batch of square matrices.
+    """
+    n = bmat.size(-1)
+    dims = torch.arange(n, out=bmat.new(n)).long()
+    if isinstance(dims, Variable):
+        dims = dims.data # TODO: why can't I index with a Variable?
+    return bmat[...,dims,dims]
+
+
+def _batch_mahalanobis(L, x):
+    r"""
+    Computes the squared Mahalanobis distance :math:`\mathbf{x}^\top\mathbf{M}^{-1}\mathbf{x}` 
+    for a factored :math:`\mathbf{M} = \mathbf{L}\mathbf{L}^\top`. 
+    
+    Accepts batches for both L and x.
+    """
+    # TODO: use `torch.potrs` or similar once a backwards pass is implemented.
+    flat_L = L.unsqueeze(0).contiguous().view((-1,)+L.shape[-2:])
+    L_inv = torch.stack([torch.inverse(Li.t()) for Li in flat_L]).view(L.shape)
+    return (x.unsqueeze(-1) * L_inv).sum(-2).pow(2.0).sum(-1)
+
+
+class MultivariateNormal(Distribution):
+    r"""
+    Creates a multivariate normal (also called Gaussian) distribution
+    parameterized by a mean vector and a covariance matrix.
+    
+    The multivariate normal distribution can be parameterized either 
+    in terms of a positive definite covariance matrix :math:`\mathbf{\Sigma}` 
+    or a lower-triangular matrix :math:`\mathbf{L}` such that 
+    :math:`\mathbf{\Sigma} = \mathbf{L}\mathbf{L}^\top` as obtained via e.g. 
+    Cholesky decomposition of the covariance.
+
+    Example:
+
+        >>> m = MultivariateNormal(torch.zeros(2), torch.eye(2))
+        >>> m.sample()  # normally distributed with mean=`[0,0]` and covariance_matrix=`I`
+        -0.2102
+        -0.5429
+        [torch.FloatTensor of size 2]
+
+    Args:
+        loc (Tensor or Variable): mean of the distribution
+        covariance_matrix (Tensor or Variable): positive-definite covariance matrix
+        scale_tril (Tensor or Variable): lower-triangular factor of covariance
```

Comment:
It looks like we're using `scale_tril` to implement pretty much all of the operations here, so it would be nice to add a note in the docs, saying that it's preferable to provide this arg.